### PR TITLE
Getbytes encoding fix

### LINKED ISSUE #105: getBytes encoding should be explicit
Extending discussion in #95 there are many places in the project source where platform dependent character encoding is used via a call to `getBytes()` instead of for instance `getBytes(utf8)` (or whatever is appropriate). All of these calls should use an explicit `Charset`.

A quick grep showed these:

```
./src/main/java/com/microsoft/graph/http/DefaultHttpProvider.java:                    bytesToWrite = serializeObject.getBytes();
./src/main/java/com/microsoft/graph/http/DefaultHttpProvider.java:        InputStream in = new ByteArrayInputStream("{}".getBytes());
./src/main/java/com/microsoft/graph/models/extensions/Multipart.java:		addPart(name, contentType, content.getBytes());
./src/main/java/com/microsoft/graph/models/extensions/Multipart.java:		out.write(partContent.getBytes());
./src/main/java/com/microsoft/graph/models/extensions/Multipart.java:		out.write(returnContent.getBytes());
./src/main/java/com/microsoft/graph/models/extensions/Multipart.java:		finalStream.write(addEnding().getBytes());
./src/test/java/com/microsoft/graph/http/MockConnection.java:        return new ByteArrayInputStream(mData.getJsonResponse().getBytes());
./src/test/java/com/microsoft/graph/functional/SharePointTests.java://        String url = "u!" + Base64.getUrlEncoder().encodeToString(testSite.webUrl.getBytes());
./src/test/java/com/microsoft/graph/functional/OneNoteTests.java:        byte[] pageStream = content.getBytes();
```

### PR DESCRIPTION
**Pull request by NakulSabharwal**
Fixes #105 

### Changes proposed in this pull request
- Set getbytes in Multipart to US-ASCII.
- Set getbytes in jsonResponse to UTF-8.
- Added getter and setter for boundary and  Content-Type of Multipart.
- Removed function addPart(String name, String contentType, String content).
- Renamed function addPart(String name, String contentType, byte[] byteArray) To addFormData(String name, String contentType, byte[] byteArray).
- Added function addPart(Map<String, String> headers, byte[] content) for user provided headers and byte content.

### Other links
-https://tools.ietf.org/html/rfc7578

### CODE REVIEW

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java [hunk 1]
```diff
@@ -33,46 +38,132 @@ public Multipart() {
...
+	
+	private void writePartData(String partContent, byte[] byteArray) throws IOException{
+		out.write(partContent.getBytes(MULTIPART_ENCODING));
+		out.write(byteArray);
+		String returnContent = RETURN + RETURN;
+		out.write(returnContent.getBytes(MULTIPART_ENCODING));
+	}
+	
+	private String createPartHeader(Map<String, String> headers, String name, String contentType, String filename) {
+		String partContent = addBoundary();
```
Comment (davidmoten) [inline src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: Nit: make `partContent` a `StringBuilder` to avoid unnecessary copying

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java [hunk 2]
```diff
@@ -33,46 +38,132 @@ public Multipart() {
...
+	}
+	
+	private void writePartData(String partContent, byte[] byteArray) throws IOException{
+		out.write(partContent.getBytes(MULTIPART_ENCODING));
+		out.write(byteArray);
+		String returnContent = RETURN + RETURN;
+		out.write(returnContent.getBytes(MULTIPART_ENCODING));
+	}
+	
+	private String createPartHeader(Map<String, String> headers, String name, String contentType, String filename) {
```
Comment (davidmoten) [inline src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: Make this method static if doesn't use parent object state

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java [hunk 3]
```diff
@@ -33,46 +38,132 @@ public Multipart() {
...
+			List<String> mainHeaders = Arrays.asList("Content-Disposition","Content-Type","charset");
+			
+			if(headers.containsKey("Content-Disposition")) {			
+				partContent += "Content-Disposition:"+headers.get("Content-Disposition")+";"; 
+				for (Map.Entry<String,String> entry : headers.entrySet()) {
+					if(listContentDisposition.contains(entry.getKey())) {
+						partContent += " " + entry.getKey() + "=\"" + entry.getValue() + "\";";
+					}
+				}
+				partContent = partContent.substring(0, partContent.length()-1);
```
Comment (davidmoten) [inline src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: You want to strip the final semicolon. Why's that?

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java (hunk 2 above)
Comment (davidmoten) [reply src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: Please add unit test coverage of this method.

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java (hunk 3 above)
Comment (darrelmiller) [reply src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: Content-Disposition parameters are prefixed by semi-colon, so removing the trailing one is technically correct.

However, I'm a little puzzled as to this approach where we pretend Content-Disposition parameters are headers and then we lift them out of the headers collection to construct the Content-Disposition header value.  The list of headers should just be a list of HTTP headers where the content-disposition value is pre-constructed.  If we need a helper to build that value, then let's build a createContentDispositionValue() function.

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java [hunk 4]
```diff
@@ -82,7 +173,7 @@ public void addPart(String name, String contentType, byte[] byteArray) throws IO
 	 * @throws IOException Throws an exception if the output stream cannot be written to
 	 */
 	public void addHtmlPart(String name, String content) throws IOException {
-		addPart(name, "text/html", content);
+		addFormData(name, "text/html", content.getBytes(MULTIPART_ENCODING));
```
Comment (darrelmiller) [inline src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: We can't make assumptions about the encoding of HTML content.  If we keep this addHtmlPart method, it needs to accept a byte array and not a string.

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java [hunk 5]
```diff
@@ -33,46 +38,132 @@ public Multipart() {
...
+			for(Map.Entry<String,String> entry : headers.entrySet()) {
+				if(mainHeaders.contains(entry.getKey())==false && listContentDisposition.contains(entry.getKey())==false) {
+					partContent += entry.getKey() +":"+entry.getValue() + RETURN;
+				}
+			}
+			partContent += RETURN;
+		}
+		else if(filename != null && name != null) {
+			partContent +=
+					"Content-Disposition:form-data; name=\"" + name + "\"" + "; filename=\"" + filename + "\"" + RETURN +
```
Comment (darrelmiller) [inline src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: There needs to be a space after the colon in the Content-Disposition header.

File: src/test/java/com/microsoft/graph/functional/OneNoteTests.java [hunk 6]
```diff
@@ -429,6 +434,59 @@ public void testMultipartPost(){
             fail("Unable to write to output stream");
         }
     }
+    
+    /**
+     * Test posting multipart content to a page
+     */
+    @Test
+    public void testMultipartHeadersMapPost(){
```
Comment (deepak2016) [inline src/test/java/com/microsoft/graph/functional/OneNoteTests.java]: rename to testMultipartPostWithHeadersMap

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java (hunk 1 above)
Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: Using String.format(...) now for building String

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java (hunk 2 above)
Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: This method cannot be made static because it uses boundary string which is dependent on parent object state.

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java (hunk 3 above)
Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: - I have now changed the way in developer will use this addPart(Map<String, String> headers, byte[] content) and added Helper function String createContentHeaderValue(String contentValue, Map<String, String> contentDispParameter). 
- contentDispParameter map will contain (key=parameterName, Value=parameterValue). eg. (key="name", value="name-val").
- createContentHeaderValue helper function will construct the string "contentValue;name="nameval";..." which will be then passed in headers map as (key=Content-Disposition, value=string from helper function), (key=Content-Type, value=string from helper).

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java (hunk 4 above)
Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: Changed addHtmlPart to accept bytes.

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java (hunk 5 above)
Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: Added space after the colon in the Content-Disposition header.

File: src/test/java/com/microsoft/graph/functional/OneNoteTests.java (hunk 6 above)
Comment (NakulSabharwal) [reply src/test/java/com/microsoft/graph/functional/OneNoteTests.java]: Renamed unit test function to testMultipartPostWithHeadersMap.

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java [hunk 7]
```diff
@@ -33,56 +36,158 @@ public Multipart() {
...
+	
+	private void writePartData(String partContent, byte[] byteArray) throws IOException{
+		out.write(partContent.getBytes(MULTIPART_ENCODING));
+		out.write(byteArray);
+		String returnContent = RETURN + RETURN;
+		out.write(returnContent.getBytes(MULTIPART_ENCODING));
+	}
+	
+	private String createPartHeader(String name, String contentType, String filename) {
+        String partContent = addBoundary();
```
Comment (davidmoten) [inline src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: if you are reworking this use a StringBuilder.

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java [hunk 8]
```diff
@@ -33,56 +36,158 @@ public Multipart() {
...
+        if(filename != null && name != null) {
```
Comment (davidmoten) [inline src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: Though the if statements may be correct you get more clarity for reading (and stop double checking values) by nesting like so:

```java
if (filename !=null) {
    if (name != null) {
        blah;
    } else {
        blah;
    }
} else if (name!=null) {
   blah;
} else {
   blah;
} 
```

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java [hunk 9]
```diff
@@ -33,56 +36,158 @@ public Multipart() {
...
+        	String partContentWithNameAndFilename = "Content-Disposition: form-data; name=\"%s\"; filename=\"%s\"" + RETURN + "Content-Type:%s" + RETURN + RETURN;
```
Comment (davidmoten) [inline src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: Are you missing a space again after ContentType:? More occurrences below.

File: src/test/java/com/microsoft/graph/functional/OneNoteTests.java [hunk 10]
```diff
@@ -429,6 +434,59 @@ public void testMultipartPost(){
             fail("Unable to write to output stream");
         }
     }
+    
+    /**
+     * Test posting multipart content to a page
+     */
+    @Test
+    public void testMultipartPostWithHeadersMap(){
```
Comment (davidmoten) [inline src/test/java/com/microsoft/graph/functional/OneNoteTests.java]: This is useful but what I'd like to see is coverage of the `createPartHeader` method which has a few conditional branches. If you wrote those simple tests you would see the formatting problems yourself and they would not need review iterations. Because you didn't explicilty add a test for `createPartHeader` I'm going to run through what I would expect to see for testing of a method like it:

* make `createPartHeader` static (including whatever state was used in its parameters)
* give the method default visibility
* add the annotation @VisibleForTesting or add comment // VisibleForTesting
* write numerous small test methods that cover every branch and condition of `createPartHeader` only (15 minutes of work to knock up)
* Keep the unit tests low level, only testing `createPartHeader` 

@darrelmiller I know we have discussed briefly how much unit testing to do but this is an obvious easy win and easy to write. I'm happy for testing paths not to be coverered but it should be deliberate and argued as to why it is not being done. 

On closer inspection the current test method performs no useful assertions so gives us no coverage at all of the changes made!

File: src/test/java/com/microsoft/graph/functional/OneNoteTests.java [hunk 11]
```diff
@@ -429,6 +434,59 @@ public void testMultipartPost(){
...
+            options.add(multipart.header());
+            
+            // Post the multipart content
+            OnenotePage page = orb
+            		.sections(testSection.id)
+            		.pages()
+            		.buildRequest(options)
+            		.post(multipart.content());
+            assertNotNull(page);
+        } catch (Exception e) {
```
Comment (davidmoten) [inline src/test/java/com/microsoft/graph/functional/OneNoteTests.java]: try catch unnecessary, just add `throws Exception` to the method. At the moment when fails we wouldn't even get a stack trace to help us.

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java (hunk 7 above)
Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: Used StringBuilder now for part header's value-parameter.

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java (hunk 8 above)
Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: Restructured the conditional statements.

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java (hunk 9 above)
Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: Corrected it with space.

File: src/test/java/com/microsoft/graph/functional/OneNoteTests.java (hunk 10 above)
Comment (NakulSabharwal) [reply src/test/java/com/microsoft/graph/functional/OneNoteTests.java]: - Wrote unit tests for createPartHeader. 
- createPartHeader cannot be made static because it uses addBoundary function which uses Boundary string that can be different for different multiparts. Suppose, developer wants to use Multipart at multiple places in application at those places the boundary value will be different so it cannot be made static, in turn then addBoundary and createPartHeader cannot be made static because of usage of non-static boundary member.

File: src/test/java/com/microsoft/graph/functional/OneNoteTests.java (hunk 11 above)
Comment (NakulSabharwal) [reply src/test/java/com/microsoft/graph/functional/OneNoteTests.java]: Changed it to throw Exception.

**Comment by deepak2016** [comment]
I would like to complete this PR as soon as possible, so that we can refresh SDK on Monday. @davidmoten @darrelmiller kindly review.

File: src/test/java/com/microsoft/graph/functional/OneNoteTests.java [hunk 12]
```diff
@@ -429,6 +434,55 @@ public void testMultipartPost(){
...
+    	List<Option> options = new ArrayList<Option>();
+    	options.add(multipart.header());
+
+    	// Post the multipart content
+    	OnenotePage page = orb
+    			.sections(testSection.id)
+    			.pages()
+    			.buildRequest(options)
+    			.post(multipart.content());
+    	assertNotNull(page);
```
[the remark this answers is not in the document]
Comment (NakulSabharwal) [reply src/test/java/com/microsoft/graph/functional/OneNoteTests.java]: The assertions should be made for checking notNull or comparison with expected url, request headers, json?

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java [hunk 13]
```diff
@@ -33,56 +37,160 @@ public Multipart() {
...
+	 * @param contentValue The content header value
+	 * @param contentDispParameter Map containing content paramter's key and value pair 
+	 * @return content header value and parameter string 
+	 */
+	public static String createContentHeaderValue(String contentValue, Map<String, String> contentDispParameter) {
+		String contentHeaderValue = contentValue;
+		
+		if(contentDispParameter != null) {
+			for(Map.Entry<String,String> entry : contentDispParameter.entrySet())
+	            contentHeaderValue += ";" + entry.getKey() + "=\"" + entry.getValue() + "\"";
```
[the remark this answers is not in the document]
Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: indented now. Although IDE was showing indented earlier.

File: src/test/java/com/microsoft/graph/functional/OneNoteTests.java [hunk 14]
```diff
@@ -429,6 +437,65 @@ public void testMultipartPost(){
...
+    	assertEquals(expectedRequestUrl, pageReq.getRequestUrl());
+    	IOnenotePageCollectionRequest request = pageReq.buildRequest(options);
+    	assertNotNull(request);
+    	OnenotePageCollectionRequest pageCollectionReq = (OnenotePageCollectionRequest)request;
+    	List<HeaderOption> headeroption = pageCollectionReq.getHeaders();
+    	assertEquals("Content-Type", headeroption.get(0).getName());
+    	String expectedHeaderValue = "multipart/form-data; boundary=\""+multipart.getBoundary()+"\"";
+    	assertEquals(expectedHeaderValue, headeroption.get(0).getValue().toString());
+    	assertNotNull(multipart.content());
+    	OnenotePage page = pageReq.buildRequest(options)
```
Comment (deepak2016) [inline src/test/java/com/microsoft/graph/functional/OneNoteTests.java]: you have already taken out request object in line 487, why making pageReq.buildRequest(options) call again?

Comment (NakulSabharwal) [reply src/test/java/com/microsoft/graph/functional/OneNoteTests.java]: changed it to use request.post(...)

File: src/main/java/com/microsoft/graph/models/extensions/Multipart.java [hunk 15]
```diff
@@ -33,56 +37,160 @@ public Multipart() {
...
+        return partContent.toString();
+	}
+	
+	/**
+	 * Create content headers value and parameter
+	 * @param contentValue The content header value
+	 * @param contentDispParameter Map containing content paramter's key and value pair 
+	 * @return content header value and parameter string 
+	 */
+	public static String createContentHeaderValue(String contentValue, Map<String, String> contentDispParameter) {
```
Comment (darrelmiller) [inline src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: I would just drop this method.  There isn't any commonality between Content-Type and Content-Disposition values.  We can build helper methods for creating header values that are semi-colon delimited parameter lists at a later point that can be used in other places.  This method just creates confusion and we are not using it except for in tests.

Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/models/extensions/Multipart.java]: The content header's value and parameter string building is common in Content-Type and Content-Disposition. eg. 
Content-Disposition:form-data; name="someName"
Content-Type: text/html; charset="someVal"
For this common structure of building I used one helper function for both headers.

**Comment by NakulSabharwal** [comment]
@davidmoten As all major comments stand resolved, we are going ahead and merging this PR, kindly open a github issue if you see any issue with the code and we will take that as a separate PR/issue.
